modules/imgcodecs/src/loadsave.cpp

```diff
@@ -1032,6 +1035,222 @@ bool haveImageWriter( const String& filename )
     return !encoder.empty();
 }
 
+class ImageCollection::IImageCollection {
+public:
+    IImageCollection() = default;
+    IImageCollection(const std::string&  filename, int flags);
+    void setup(const String& img, int flags);
+    size_t size() const;
+    Mat read();
+    int width() const;
+    int height() const;
+    bool readHeader();
+    Mat readData();
+    bool advance();
+    int currentIndex() const;
+    void reinitDecoder();
+
+private:
+    String m_filename;
+    int m_flags{};
+    int m_size{};
+    int m_width{};
+    int m_height{};
+    int m_current{};
+    ImageDecoder m_decoder;
+};
+
+ImageCollection::IImageCollection::IImageCollection(std::string const& filename, int flags) {
+    this->setup(filename, flags);
+}
+
+void ImageCollection::IImageCollection::setup(String const& filename, int flags) {
+    m_filename = filename;
+    m_flags = flags;
+
+#ifdef HAVE_GDAL
+    if (m_flags != IMREAD_UNCHANGED && (m_flags & IMREAD_LOAD_GDAL) == IMREAD_LOAD_GDAL) {
+        m_decoder = GdalDecoder().newDecoder();
+    }
+    else {
+#endif
+    m_decoder = findDecoder(filename);
+#ifdef HAVE_GDAL
+    }
+#endif
+
+
+    if (!m_decoder) {
+        throw cv::Exception(1, "No Encoder found for the file named:" + m_filename,
+                            String("ImageCollection::Constructor"),
+                            __FILE__,
+                            __LINE__);
+    }
+
+    m_decoder->setSource(filename);
+
+    try {
+
+        if (!m_decoder->readHeader())
+            throw cv::Exception(1, "Can not read image header in ImageCollection" +
+                                String(" file named " + m_filename),
+                                String("ImageCollection::Constructor"),
+                                __FILE__,
+                                __LINE__);
+    }
+    catch (const cv::Exception& e)
+    {
+        std::cerr << "ImageCollection class:: can't read header: " << e.what() << std::endl << std::flush;
+        throw cv::Exception();
+    }
+    catch (...)
+    {
+        std::cerr << "ImageCollection class:: can't read header: unknown exception" << std::endl << std::flush;
+        throw cv::Exception();
+    }
+
+    size_t count = 1;
+    while(m_decoder->nextPage()) count++;
+
+    m_size = count;
+    reinitDecoder();
+}
+
+size_t ImageCollection::IImageCollection::size() const { return m_size; }
```

Comment:
This returns the number of pages in image. I think i should find a new, better name for this one. So we can return the current page Size(width, height).